src/flp/szk.rs

```diff
@@ -57,29 +60,144 @@ pub enum SzkError {
 
 /// Contains an FLP proof share, and if joint randomness is needed, the blind
 /// used to derive it and the other party's joint randomness part.
-#[derive(Clone)]
-pub enum SzkProofShare<F: FftFriendlyFieldElement, const SEED_SIZE: usize> {
-    /// Leader's proof share is uncompressed. The first Seed is a blind, second
-    /// is a joint randomness part.
+#[derive(Debug, Clone)]
+pub enum SzkProofShare<F, const SEED_SIZE: usize> {
+    /// Leader's proof share is uncompressed.
     Leader {
+        /// Share of an FLP proof, as a vector of Field elements.
         uncompressed_proof_share: Vec<F>,
-        leader_blind_and_helper_joint_rand_part: Option<(Seed<SEED_SIZE>, Seed<SEED_SIZE>)>,
+        /// Set only if joint randomness is needed. The first Seed is a blind, second
+        /// is the helper's joint randomness part.
+        leader_blind_and_helper_joint_rand_part_opt: Option<(Seed<SEED_SIZE>, Seed<SEED_SIZE>)>,
     },
     /// The Helper uses one seed for both its compressed proof share and as the blind for its joint
     /// randomness.
     Helper {
+        /// The Seed that acts both as the compressed proof share and, optionally, as the blind.
         proof_share_seed_and_blind: Seed<SEED_SIZE>,
-        leader_joint_rand_part: Option<Seed<SEED_SIZE>>,
+        /// The leader's joint randomness part, if needed.
+        leader_joint_rand_part_opt: Option<Seed<SEED_SIZE>>,
     },
 }
 
+impl<F: FieldElement + Decode, const SEED_SIZE: usize> ParameterizedDecode<(usize, usize, bool)>
+    for SzkProofShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        (role, bits, is_random): &(usize, usize, bool),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match role {
+            0 => Ok(SzkProofShare::Leader {
+                uncompressed_proof_share: Vec::<F>::decode_with_param(bits, bytes)?,
+                leader_blind_and_helper_joint_rand_part_opt: is_random.then_some((
+                    Seed::<SEED_SIZE>::decode(bytes)?,
+                    Seed::<SEED_SIZE>::decode(bytes)?,
+                )),
+            }),
+            1 => Ok(SzkProofShare::Helper {
+                proof_share_seed_and_blind: Seed::<SEED_SIZE>::decode(bytes)?,
+                leader_joint_rand_part_opt: is_random.then_some(Seed::<SEED_SIZE>::decode(bytes)?),
+            }),
+            _ => Err(CodecError::UnexpectedValue),
```

Comment:
`UnexpectedValue` is really intended for malformed messages, not the decoding parameter. Perhaps replace `role: usize` with `is_leader: bool`, which would eliminate this case.